### PR TITLE
don't let UNIQUE and PRIMARY KEY be seen as column names, skip internal sqlite tables

### PR DESCRIPTION
**Pull request by adamdecaf**
These are useful as they're slightly more complex (and real-world) sqlite dbs. 

Here's the tables and data contained. Both are from a fresh install of the browser.

<details>
<summary>Chrome</summary>

```
$ sqlite3 browser/chrome/testdata/History 
SQLite version 3.19.3 2017-06-27 16:48:08
Enter ".help" for usage hints.
sqlite> .tables
downloads                meta                     urls                   
downloads_slices         segment_usage            visit_source           
downloads_url_chains     segments                 visits                 
keyword_search_terms     typed_url_sync_metadata
sqlite> select count(*) from downloads;
0
sqlite> select count(*) from downloads_slices;
0
sqlite> select count(*) from downloads_url_chains;
0
sqlite> select count(*) from keyword_search_terms;
0
sqlite> select count(*) from meta;
3
sqlite> select * from meta;
mmap_status|-1
version|36
last_compatible_version|16
sqlite> select * from segment_usage;
1|1|13156408800000000|1
sqlite> select * from segments;
1|http://google.com/|1
sqlite> select * from typed_url_sync_metadata;
sqlite> select * from urls;
1|http://google.com/|Google|1|1|13156455493534189|0
2|http://www.google.com/|Google|1|0|13156455493534189|0
3|https://www.google.com/|Google|1|0|13156455493534189|0
sqlite> select * from visit_source;
sqlite> select * from visits;
1|1|13156455493534189|0|268435457|1|0
2|2|13156455493534189|1|2147483649|0|0
3|3|13156455493534189|2|2684354561|0|1039690
```

</details>

<details>
<summary>Firefox</summary>

```
$ sqlite3 browser/firefox/testdata/places.sqlite
SQLite version 3.19.3 2017-06-27 16:48:08
Enter ".help" for usage hints.
sqlite> .tables
moz_anno_attributes    moz_historyvisits      moz_keywords         
moz_annos              moz_hosts              moz_places           
moz_bookmarks          moz_inputhistory     
moz_bookmarks_deleted  moz_items_annos      
sqlite> select * from moz_anno_attributes;
1|mobile/bookmarksRoot
2|bookmarkProperties/description
3|Places/SmartBookmark
sqlite> select * from moz_annos;
sqlite> select * from moz_bookmarks;
1|2||0|0||||1511549555759000|1511549555976000|root________|1|1
2|2||1|0|Bookmarks Menu|||1511549555759000|1511549555976000|menu________|1|5
3|2||1|1|Bookmarks Toolbar|||1511549555759000|1511549555923000|toolbar_____|1|5
4|2||1|2|Tags|||1511549555759000|1511549555759000|tags________|1|1
5|2||1|3|Other Bookmarks|||1511549555759000|1511549555894000|unfiled_____|1|2
6|2||1|4|mobile|||1511549555759000|1511549555759000|mobile______|1|1
7|1|2|3|1|Getting Started|||1511549555897000|1511549555900000|vaOSWVBus5Xb|0|2
8|2||2|2|Mozilla Firefox|||1511549555900000|1511549555905000|mbH-q0s0dRck|0|5
9|1|3|8|0|Help and Tutorials|||1511549555901000|1511549555902000|Yk4kmW0ggs-g|0|2
10|1|4|8|1|Customize Firefox|||1511549555902000|1511549555903000|2Tqd-7CX_4IK|0|2
11|1|5|8|2|Get Involved|||1511549555904000|1511549555905000|PZ5ryma0-vic|0|2
12|1|6|8|3|About Us|||1511549555905000|1511549555906000|IrZIt1HsbFEl|0|2
13|1|7|3|0|Most Visited|||1511549555923000|1511549555955000|lWpPgrVeRjwD|1|2
14|1|8|2|0|Recent Tags|||1511549555955000|1511549555969000|Gyq4voQ9v5Wk|1|2
15|3||2|1||||1511549555976000|1511549555976000|oZKWNKmvfGcz|1|1
sqlite> select * from moz_bookmarks_deleted;
sqlite> select * from moz_historyvisits;
1|0|1|1511549555740045|1|0
2|0|9|1511549555969726|1|0
3|2|10|1511549556034053|5|0
sqlite> select * from moz_hosts;
1|mozilla.org|140|0|
2|support.mozilla.org|140|0|
sqlite> select * from moz_inputhistory;
sqlite> select * from moz_items_annos;
1|6|1|1|0|4|1|240441240|240441240
2|3|2|Add bookmarks to this folder to see them displayed on the Bookmarks Toolbar|0|4|3|1511549555896000|1511549555896000
3|13|3|MostVisited|0|4|3|1511549555955000|1511549555955000
4|14|3|RecentTags|0|4|3|1511549555969000|1511549555969000
sqlite> select * from moz_keywords;
sqlite> select * from moz_places;
1|https://www.mozilla.org/en-US/firefox/57.0/firstrun/|Welcome to Firefox|gro.allizom.www.|1|0|0|100|1511549555740045|lPFu5Q-5WggG|0|47357702733362||
2|https://www.mozilla.org/en-US/firefox/central/||gro.allizom.www.|0|0|0|140||7wnKOlnK7xYW|1|47356370932282||
3|https://support.mozilla.org/en-US/products/firefox||gro.allizom.troppus.|0|0|0|140||WJEEUKJR5m_Q|1|47357795150914||
4|https://www.mozilla.org/en-US/firefox/customize/||gro.allizom.www.|0|0|0|140||GbNOnfHsBSSJ|1|47357014640010||
5|https://www.mozilla.org/en-US/contribute/||gro.allizom.www.|0|0|0|140||lwa0vuxLvxXt|1|47358034485371||
6|https://www.mozilla.org/en-US/about/||gro.allizom.www.|0|0|0|140||OtK7Z4Q-wSTR|1|47358774953055||
7|place:sort=8&maxResults=10||.|0|1|0|0||myMb14BJnvGi|1|268505095842199||
8|place:type=6&sort=14&maxResults=10||.|0|1|0|0||mMjnLa6WCX_C|1|268505606444332||
9|https://www.mozilla.org/privacy/firefox/||gro.allizom.www.|1|1|0|25|1511549555969726|lxnNte7gemMe|0|47356411089529||
10|https://www.mozilla.org/en-US/privacy/firefox/||gro.allizom.www.|1|0|0|100|1511549556034053|KA8U_nxp9AfC|0|47358032558425||
```

</details>

### CODE REVIEW

**Comment by adamdecaf** [comment]
The column name checks can be added to these after https://github.com/go-sqlite/sqlite3/pull/17

**Comment by zellyn** [comment]
lgtm

File: sqlite_test.go [hunk 1]
```diff
@@ -28,6 +28,20 @@ func TestFileOpen(t *testing.T) {
 			npages:  4,
 			pagesz:  1024,
 		},
+		{
+			// Chrome history sqlite db
+			fname: "testdata/History",
```
Comment (sbinet) [inline sqlite_test.go]: could we have a more interesting name for this file?
`chrome-history.sqlite` would work with me :)

File: sqlite_test.go [hunk 2]
```diff
@@ -28,6 +28,20 @@ func TestFileOpen(t *testing.T) {
...
+			version: 3020000,
+			npages: 28,
+			pagesz: 4096,
+		},
+		{
+			// Firefox history sqlite db
+			fname: "testdata/places.sqlite",
```
Comment (sbinet) [inline sqlite_test.go]: ditto.
`firefox-history.sqlite` ?

**Comment by adamdecaf** [comment]
Files are renamed. :)

**Comment by sbinet** [comment]
ah... I didn't notice at first, but the tests are failing.
seems real and related to the new files that trigger an invalid index (`=-1`) for column name extraction...

**Comment by adamdecaf** [comment]
Dang. It looks like #17 was too simple of a fix.

**Comment by adamdecaf** [comment]
Alright, this parses the tables now. I'm not too familiar with sqlite's sql dialect, but are there other statements which would appear?

The tests panic, because of https://github.com/go-sqlite/sqlite3/commit/b57e325f545db253ce37589fdb7073a2f1dcd545#r26970773 

Would y'all want the full schema for chrome and firefox laid out in this test? It'll make the file pretty long.

File: file.go [hunk 3]
```diff
@@ -19,6 +19,10 @@ const (
 	printfDebug = false
 )
 
+var (
+	idxtokens = []string{"UNIQUE", "PRIMARY KEY"}
```
Comment (adamdecaf) [inline file.go]: There's probably a better name than `idxtoken` here..

File: file.go [hunk 4]
```diff
@@ -204,6 +213,21 @@ func (db *DbFile) init() error {
...
+		for i := range parts {
+			if i >= len(parts) {
+				break // we removed at least one elem, so avoid out of bounds read
+			}
+
+			parts[i] = strings.TrimSpace(parts[i])
+			for j := range idxtokens {
+				if strings.HasPrefix(parts[i], idxtokens[j]) {
+					// drop all other elements
+					parts = parts[:i]
```
Comment (adamdecaf) [inline file.go]: Is it safe to assume `UNIQUE` and/or `PRIMARY KEY` are always the last things printed in a table definition?

**Comment by zellyn** [comment]
Don't need full schemas laid out in the test… just make sure the test that loads those example tables tries to parse the full table definitions. That way it'll catch errors if it's broken.

**Comment by adamdecaf** [comment]
https://github.com/go-sqlite/sqlite3/pull/18/commits/99dceb84f5fa8361dfaa53a0b4fe503d7ee9d208 checks the parsed table counts.

File: file.go (hunk 4 above)
Comment (zellyn) [reply file.go]: No idea on the guarantees. The grammar would be the authoritative reference, and is pretty clear…

Comment (adamdecaf) [reply file.go]: Yea. I added the other table constraints from https://sqlite.org/lang_createtable.html

**Pull request merged**